core/src/smartcontracts/isi/permissions.rs

```diff
@@ -15,35 +15,88 @@ use super::{judge::Judge, *};
 pub fn check_instruction_permissions(
     account_id: &AccountId,
     instruction: &Instruction,
-    instruction_judge: &dyn Judge<Operation = Instruction>,
-    query_judge: &dyn Judge<Operation = QueryBox>,
     wsv: &WorldStateView,
 ) -> std::result::Result<(), TransactionRejectionReason> {
-    let granted_instructions = &super::roles::unpack_if_role_grant(instruction.clone(), wsv)
+    let granted_instructions = &unpack_if_role_grant(instruction.clone(), wsv)
         .expect("Infallible. Evaluations have been checked by instruction execution.");
-    check_permissions_directly(account_id, granted_instructions, instruction_judge, wsv)?;
+    check_permissions_directly(account_id, granted_instructions, wsv)?;
 
-    let revoked_instructions = &super::roles::unpack_if_role_revoke(instruction.clone(), wsv)
+    let revoked_instructions = &unpack_if_role_revoke(instruction.clone(), wsv)
         .expect("Infallible. Evaluations have been checked by instruction execution.");
-    check_permissions_directly(account_id, revoked_instructions, instruction_judge, wsv)?;
+    check_permissions_directly(account_id, revoked_instructions, wsv)?;
 
-    check_query_in_instruction(account_id, instruction, wsv, query_judge)
-        .map_err(|reason| NotPermittedFail { reason })
+    check_permission_recursively(account_id, instruction, wsv)?;
```

Comment:
I would bet this ends up validating simple instruction twice. Am I wrong?

Reply:
These type of mistakes happen when you process an item (in this case instruction) in multiple places instead of doing it in one place